features/filesystem/storagelite/source/StorageLiteFS.cpp

```diff
@@ -0,0 +1,320 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "StorageLiteFS.h"
+#include "mbed_retarget.h"
+#include "string.h"
+#include "stdlib.h"
+
+typedef struct {
+    char *file_name;
+    int open_flags;
+} file_handle_data_t;
+
+// Convert basic StorageLite error numbers to errno based ones
+static int convert_to_errno(int err)
+{
+    switch (err) {
+        case STORAGELITE_SUCCESS:
+            return  0;
+        case STORAGELITE_READ_ERROR:
+            return -EIO;
+        case STORAGELITE_WRITE_ERROR:
+            return -EIO;
+        case STORAGELITE_NOT_FOUND:
+            return -ENOENT;
+        case STORAGELITE_DATA_CORRUPT:
+            return -EILSEQ;
+        case STORAGELITE_BAD_VALUE:
+            return -EINVAL;
+        case STORAGELITE_BUFF_TOO_SMALL:
+            return -ENOMEM;
+        case STORAGELITE_NO_SPACE_ON_BD:
+            return -ENOSPC;
+        case STORAGELITE_OS_ERROR:
+            return -EINTR; // FIXME: Find a better error
+        case STORAGELITE_UNINITIALIZED:
+            return -ENXIO; // FIXME: Find a better error
+        case STORAGELITE_MAX_FILES_REACHED:
+            return -ENFILE;
+        case STORAGELITE_NOT_SUPPORTED:
+            return -ENOSYS;
+        default:
+            return -err;
+    }
+}
+
+StorageLiteFS::StorageLiteFS(const char *name, StorageLite *stlite, uint32_t flags)
+    : FileSystem(name), _stlite(stlite), _flags(flags), _bd(0)
+{
+    MBED_ASSERT(stlite);
+    mount(_bd);
+}
+
+StorageLiteFS::~StorageLiteFS()
+{
+    unmount();
+}
+
+int StorageLiteFS::mount(BlockDevice *bd)
+{
+    _bd = bd;
```

Comment:
Error handling/checking of input parameters must be done, what if bd is NULL?

Reply:
We don't check NULL pointers elsewhere. Worst case, code will hard fault, and the NULL dereference can easily be found with a debugger.

Reply:
Yup, and it’s not good. Recently the code base I am touching I try to add where ever I can, hence pointed out. Can be found with debugger but not good for end product

Reply:
Interesting, I think we disagree here. Should we also be asserting on misaligned pointers?

Reply:
Error handling does not mean only asserting, but to actually avoid sudden crash and respond with proper errror message/code to help application recover (if possible). Misaligned address access depends on type of underlying block device. You have a good point, I will check how SD driver behaves as I don’t remember on top of my mind. But device should terminate gracefully even in case of unaligned access with error and not crash suddenly.

Reply:
Strongly agree with @geky here. If you're going to check NULL pointers, you logically need to be checking for misaligned pointers, stale pointers, out-of-range pointers, corrupted data structures, so on and so forth.

If you are going to worry about the validity of pointers you need to be writing in something other than C or C++03.

Are you going to add `if (this == NULL)` checks to the top of every method too? You would logically have to in the model you advocate.

And what happens in practice if you do return an error code on an invalid pointer? The caller probably ignores it. So we get subtle effects like we've had with some `Mutex::lock()`s not working without anyone realising, causing hard to track-down problems.

Having said all that...

> the NULL dereference can easily be found with a debugger.

pyOCD still can't show a backtrace on a Cortex-M HardFault, as far as I know. It's been like that for a few years now. Rather undermines our argument, if pyOCD is our preferred debugger.

Reply:
> pyOCD still can't show a backtrace on a Cortex-M HardFault

This would be a good thing to fix!